[R&D] feat: add object storage cleaner script and testing

Rakanhf opened the pull request:
Currently QFieldCloud keeps all project files using the non-legacy storage even if they are deleted by the user. This is due to enabled versioning on Exoscale and the lack of Exoscale mechanism to delete files after certain amount of time after they are deleted. This causes excessive file storage and therefore costs.

- Add `storage_cleaner.py` for scanning and deleting logically deleted objects in S3 compatable storages
- Add `test.py` for testing the script
- Add a `README.MD`

---
A small CLI to **analyze and clean logically deleted objects** in **versioned S3-compatible buckets**.

This script is **provider-agnostic** works with any S3-compatible service that supports versioning such as `MinIO`, etc.

It helps you to the following:
1. Detect whether a bucket has versioning configured.
2. Find keys whose **latest version is a delete marker** (logically deleted).
3. Report how much storage their historical versions consume.
4. **Deleting all versions** for those keys to free space.

#### Requirements
- boto3
- pytest (just for testing)

#### Usage

##### 1. No action provided or `--info`
```
python storage_cleaner.py my-versioned-bucket-name
```
```
No action specified. Use --info, --scan, or --prune.
Use --help for usage information.
Defaulting to --info.

Bucket Information:
--------------------------------------------------------------------------------
Bucket:            my-versioned-bucket-name
Versioning Status: Enabled
Region:            us-east-1
Endpoint URL:      http://localhost:8009
Prefix Filter:     No prefix specified
Deleted After:     No deleted after date specified
--------------------------------------------------------------------------------
✓ Connection successful.
✓ Bucket is accessible.

Use --scan to find logically deleted objects.
```

##### 1. Scanning for logically deleted objects
```
python storage_cleaner.py my-versioned-bucket-name --scan
```
##### 1. Deleting logically deleted objects
```
python storage_cleaner.py my-versioned-bucket-name --prune
```

#### Arguments
1. `--noinput` Skip confirmation prompt (use with --purge for automation)
2. `--deleted-after` Only process objects deleted after this date (ISO 8601 format, e.g. 2024-12-01:00:00:00) Cannot be used with `--deleted-since.`
3. `--deleted-since` Time duration ago (e.g. '3 days', '2 weeks'). Cannot be used with `--deleted-after.`
4. `--prefix` Key prefix to scan
5. `--profile` Object storage profile
6. `--debug` Enable debug logging
7. `--log-file` Write a rotating logbook file (e.g. /tmp/s3-cleaner.log)
8. `--scan` Scan only
9. `--prune` Delete found logically deleted objects (requires confirmation unless --noinput)
10. `--info` Show bucket information (default behavior)

By default it will load the `default` local S3 configuration 


### Testing 
Added multiple test cases 
1. Upload files to object storage - no deleted files (shouldn't find any logically deleted objects) -> Nothing deleted
2. Upload files to object storage - delete some files (should find logically deleted objects) -> Deleted files
3. Upload files to object storage - delete some files - restore (shouldn't find any logically deleted objects) -> Nothing deleted
4. Upload files, delete some, test deleted-after filter.
11. File created, deleted, created again, deleted again. Ensure ALL versions (history) are cleaned up.


# Demo
#### Scan
![Screencast from 2025-12-25 00-35-29](https://github.com/user-attachments/assets/034de643-f837-4cac-a395-116fdb6907d1)

#### Prune
![Screencast from 2025-12-25 00-37-48](https://github.com/user-attachments/assets/1d3f758b-2e80-46fd-aa60-e8bea1e7e001)

Rakanhf committed:
feat: add object storage cleaner script and testing

Currently QFieldCloud keeps all project files using the non-legacy storage even if they are deleted by the user.
This is due to enabled versioning on Exoscale and the lack of Exoscale mechanism to delete files after certain amount of time after they are deleted.
This causes excessive file storage and therefore costs.

- Add `storage_cleaner.py` for scanning and deleting logically deleted objects in S3 compatable storages
- Add `test.py` for testing the script

gounux commented on scripts/object_storage_cleaner/storage_cleaner.py:
```diff
@@ -0,0 +1,550 @@
+        access_key_id: Optional[str] = None,
+        secret_access_key: Optional[str] = None,
+        prefix: Optional[str] = None,
+        deleted_after: Optional[datetime] = None,
```
Could we use python >= 3.9 typings, with `variable: str | None` ?

gounux commented on scripts/object_storage_cleaner/storage_cleaner.py:
```diff
@@ -0,0 +1,550 @@
+        except ValueError:
+            raise argparse.ArgumentTypeError(f"Invalid ISO 8601 date: '{value}'")
+    else:
+        return None
```
Seems that the code might produce some null results, and accepts null `value` argument as well, so maybe :

```python
def parse_dt(value: str | None) -> datetime | None:
```

?

gounux commented on scripts/object_storage_cleaner/storage_cleaner.py:
```diff
@@ -0,0 +1,550 @@
+    # Connection options
+    parser.add_argument("--prefix", help="Key prefix to scan")
+    parser.add_argument("--region", help="AWS region")
+    parser.add_argument("--profile", help="AWS profile")
```
I believe it is not necessarily `AWS`-related, S3 can also be provided by some other hosting services,

gounux reviewed (commented):
Thanks a lot, looks promising !

This is a first review round, did not checkout locally neither tested _yet_, mainly some typings & formalism comments.

Rakanhf commented on scripts/object_storage_cleaner/storage_cleaner.py:
```diff
@@ -0,0 +1,550 @@
+    ) -> Iterator[Tuple[LogicallyDeletedObject, List[VersionRef]]]:
+        """Yields logically deleted objects."""
+        current_key: Optional[str] = None
+        current_versions: List[VersionRef] = []
```
Sure I agree

Rakanhf commented on scripts/object_storage_cleaner/storage_cleaner.py:
```diff
@@ -0,0 +1,550 @@
+        help="Skip confirmation prompt (use with --purge for automation)",
+    )
+
+    args = parser.parse_args()
```
Yes, we can extract the parser function, which mainly was used in an early version of the script 

```python
def parse_args(argv: Optional[Iterable[str]] = None) -> argparse.Namespace:
    """Parse command-line arguments."""
```

But I felt that since the file size is growing I kept it in the main function, personally I don't have a strong opinion about it, both option are fine.

Rakanhf replied:
Surprisingly enough it's also mentioned In the book `Clean Code` it argues that functions must be very small, often just a few lines, and that anything beyond ~20 lines is a red flag. which I totally disagree with, since I don't think that lines of code are the only measure to judge if the code is `clean` or not. I usually prefer to have logical chunks and keeping separation, I don't personally mind a 80 lines function but does exactly two or less logical processes.

This is highly debatable 😄 , since every implementation comes with it's own restrictions. 

My thinking process is "I have two other until functions in same file = adding parser into it's own function will make file full of function to click and read" which rings a bill that Oh I need to separate to multiple files to make things more organized and easy to ready. e.g utlis.py, parser.py, main.py, cli.py, etc.

Rakanhf commented on scripts/object_storage_cleaner/storage_cleaner.py:
```diff
@@ -0,0 +1,550 @@
+    parser.add_argument("--profile", help="AWS profile")
+    parser.add_argument("--endpoint-url", help="Custom S3 endpoint URL")
+    parser.add_argument("--access-key-id", help="Access key ID")
+    parser.add_argument("--secret-access-key", help="Secret access key")
```
Mainly I kept this just as an option or even to use it with env e.g : 

```
export S3_ACCESS_KEY <>

python storage_cleaner.py --access-key-id S3_ACCESS_KEY ....... 
```

But boto3 by default looks for the default profile on `~/.aws/config` and `~/.aws/credentials` 

e.g 
```
# ~/.aws/credentials

[default]
aws_access_key_id = my_access_key
aws_secret_access_key = my_secret_access_key

# ~/.aws/config

[default]
endpoint_url = http://localhost:8009
region = us-east-1
```

Ref https://docs.aws.amazon.com/cli/v1/userguide/cli-configure-files.html#cli-configure-files-using-profiles


TL;DR : You don't need to use it you should always use profiles, but it's there for some quick access for testing / dev for those don't have aws cli setup.

suricactus reviewed (commented):
Summary of the review:

The script needs a few iterations before it gets in a better shape. So be patient :)

There are a few things that are still not addressed since the preliminary review via chat.

In short, I think the script is more complex than it should be. Both in terms of algorithmic approach, as well as way of writing it. A bit more OOP than my taste, this is a script in the end.

Note this script will be used by people who are not necessarily experts in Python. Write the script like someone who sees python for the first time will read it.

Please ping me when the current comments are addressed for a second round. Feel free to simplify things that you recognize as complex or not obvious, but were not commented in this round.

Rakanhf committed:
refactor: update type hints to use python3.9+

Rakanhf committed:
refactor: simplify datetime parsing

Rakanhf committed:
refactor: remove access key parameters from ObjectStorageScanner

Rakanhf commented on scripts/object_storage_cleaner/storage_cleaner.py:
```diff
@@ -0,0 +1,550 @@
+            return None
+
+        # Sort by time descending (latest first)
+        versions.sort(key=lambda x: x.last_modified, reverse=True)
```
`bucket.object_versions` doesn’t group by key. It just loops over what S3 returns. Since S3 sends `Versions` and `DeleteMarkers` as separate lists, keys can appear again later in the same page (like file1 → file2 → file1).


I have created a small function to test this out :

```python 
def print_buckt_order():
    # Setup
    s3 = boto3.resource("s3")
    bucket = s3.create_bucket(Bucket="demo-order")
    s3.BucketVersioning("demo-order").enable()

    # Create 1st key
    key1 = "A"
    bucket.put_object(Key=key1, Body=b"data")
    bucket.Object(key1).delete()

    # Create 2nd key
    key2 = "B"
    bucket.put_object(Key=key2, Body=b"data")
    bucket.Object(key2).delete()

    print(f"{'Key':<5} Type")
    for v in bucket.object_versions.all():
        print(f"{v.key:<5} {'DeleteMarker' if v.size is None else 'Version'}")
```

Output

```
Key   Type
A     Version
B     Version
A     DeleteMarker
B     DeleteMarker
```

which confirm that it's going to be an issue if we didn't sort them out first.

suricactus replied:
Because I am lazy, note for myself:

```
export AWS_ENDPOINT_URL=http://172.17.0.1:8009 AWS_ACCESS_KEY_ID=rustfsadmin AWS_ACCESS_SECRET_KEY=rustfsadmin
```

```
import boto3

s3 = boto3.resource("s3")
bucket_name = "qfieldcloud-local"

bucket = s3.Bucket(bucket_name)
bucket.object_versions.delete()
bucket.delete()

bucket = s3.create_bucket(Bucket=bucket_name)
s3.BucketVersioning(bucket_name).enable()

root_key = "file"
special_key = "SPECIAL__FILE__000"
bucket.put_object(Key=special_key, Body="v1")
bucket.put_object(Key=special_key, Body="v2")

for i in range(1, 1000):
    key = f"{root_key}__{i:>03d}"
    bucket.put_object(Key=key, Body=key)


bucket.Object(special_key).delete()
bucket.Object("file__998").delete()


for idx, v in enumerate(bucket.object_versions.all()):
    print(f"{idx:>03d}", v.key, v.version_id, v.size)

```

Rakanhf commented on scripts/object_storage_cleaner/storage_cleaner.py:
```diff
@@ -0,0 +1,550 @@
+
+        return status == "Enabled"
+
+    def _iter_all_versions(self) -> Iterator[VersionRef]:
```
> I still don't get the purpose of this and the VersionRef object.

Versions have a Size attribute, DeleteMarkers do not. Accessing `.size` on a Delete Marker object from the raw API would raise an error.

> I am also curious can't we abandon this whole thing and just use Bucket.list_versions? See

I'm not sure I got this right, but I couldn't find anything related to `Bucket.list_versions` am I missing something ?

suricactus replied:
You are right- there is no size attribute- but I still find

```
size = 0
if v.is_delete_marker:
  # delete markers do not have a size attribute
  size += 0
else:
  size += v["size"]
```

OR 

```
# delete markers...
size += v.get("size", 0) # might be getattr
```

This way we drop the need of two classes and a special function. Less code is more, even if we have the false perception that the classes will cahe explicit typing etc. 

Again, this is a script we must be able to debug easily by anyone, most likely people outside our team. Keep it as simple as possible. The native types from boto3 will always be better documented.

Rakanhf committed:
refactor: improve version count calculation

Rakanhf committed:
test: enhance object storage cleaner tests and fix config and assertions

- Add tests for handling large data uploads and complex version histories.

Rakanhf committed:
refactor: imrpove code readability

Rakanhf committed:
refactor: format code

Rakanhf committed:
refactor: enhance object storage cleaner functionality and output

- Enhance progress tracking by adding rogress callback
- Refactored scan and clean methods for better readability and maintainability.

Rakanhf committed:
refactor: implement logging and improve output formatting

- Add logging functionality to replace print.
- Enhanced output formatting for bucket information and progress updates.
- Introduced a TRACE logging level for detailed operation logs.
- Updated the main function to support debug logging and log file options.

Rakanhf committed:
feat: add 'since' filter for object deletion in storage cleaner

- Implemented a new command-line argument '--since' to filter deleted objects based on a specified duration.
- Added a helper function to parse various time formats for the 'since' argument.
- Enhanced tests to validate the functionality of the 'since' filter.

Rakanhf committed:
refactor: improve object storage cleaner code structure

Rakanhf committed:
test: add tests for stray delete markers and high frequency updates in object storage cleaner

Rakanhf committed:
refactor: rename ObjectStorageScanner to ObjectStorageCleaner and update references

Rakanhf committed:
refactor: rename command line arguments for clarity

suricactus commented on scripts/object_storage_cleaner/storage_cleaner.py:
```diff
@@ -443,6 +444,44 @@ def setup_logging(debug: bool = False, log_file: str | None = None) -> None:
 
+def parse_since(value: str) -> datetime:
+    """
+    Parse a duration like '3s', '30d', '2w', '3 days' and return a datetime object.
```
Do not accept the short names such as s, secs, d, w, but only the full ones. This can be disastrous for such important script, check my original comment on this highlighting the same argument.

Rakanhf replied:
Sorry for misunderstanding, I took inspiration from other places on how it's usually handled, and faced a dead end when I realized `m` can be for both minute and month 😸, so I just decided to disable months. 

But yes I agree with you about this, later on in the future we can expand or enhance it.

Rakanhf committed:
feat: add README for Object Storage Cleaner utility

Rakanhf committed:
refactor: update duration parsing in storage cleaner

- Changed regex pattern to accept full time unit words instead of abbreviations.

Rakanhf committed:
refactor: format readme document

suricactus commented on scripts/object_storage_cleaner/storage_cleaner.py:
```diff
@@ -0,0 +1,628 @@
+
+def setup_logging(debug: bool = False, log_file: str | None = None) -> None:
+    """Setup logging for the script."""
+    level = TRACE_LEVEL if log_file else (logging.DEBUG if debug else logging.INFO)
```
Please do not use inline `if` in QFieldCloud code. This is unreadable by the common folks. Do not apply such magic, make as many of the configuration required and explicit. The log level, if not by default, should be passed as a cli param/envvar and no magical logic should be applied.

suricactus commented on scripts/object_storage_cleaner/storage_cleaner.py:
```diff
@@ -0,0 +1,628 @@
+                "%(asctime)s [%(levelname)s] %(message)s", datefmt="%Y-%m-%d %H:%M:%S"
+            )
+        )
+        logger.addHandler(file_handler)
```
Leave this to the outer level, there are log rotators that are doing this much better than this code. Just kill this functionality.

suricactus replied:
One run of the script should produce one log file. We can easily do that with a timestamp suffix.

If we want cleanup or something more advanced, our good friend https://linux.die.net/man/8/logrotate exists.

suricactus commented on scripts/object_storage_cleaner/storage_cleaner.py:
```diff
@@ -0,0 +1,628 @@
+
+    def _print_progress(self, delta: int, label: str = "Deletable") -> None:
+        self._scanned_count += delta
+        if sys.stdout.isatty():
```
Do not do any branching on printing based on `tty`-ness. Just print a regular log from the caller instead. Kill this function all together.

suricactus replied:
Just print to STDOUT and you don't have to solve any of these problems. 

https://github.com/opengisch/QFieldCloud/pull/1445#discussion_r2650033258

suricactus commented on scripts/object_storage_cleaner/storage_cleaner.py:
```diff
@@ -0,0 +1,628 @@
+
+        try:
+            self.s3_client.delete_objects(
+                Bucket=self.bucket, Delete={"Objects": versions, "Quiet": True}
```
Can you please document what `Quiet` parameter means?

suricactus commented on scripts/object_storage_cleaner/storage_cleaner.py:
```diff
@@ -0,0 +1,628 @@
+    ) -> Iterator[VersionRef]:
+        """Low-level generator that yields all versions and delete markers, sorted by Key."""
+        paginator = self.s3_client.get_paginator("list_object_versions")
+        kwargs = {"Bucket": self.bucket, "MaxKeys": 10000}
```
BTW, avoid adding dicts with more than 1 key without a trailing comma.

Rakanhf committed:
fix: adjust pagination limit and format code

Rakanhf committed:
refactor: enhance code quality

suricactus commented on scripts/object_storage_cleaner/storage_cleaner.py:
```diff
@@ -0,0 +1,639 @@
+    parser.add_argument("--prefix", help="Key prefix to scan")
+    parser.add_argument("--region", help="Region")
+    parser.add_argument("--profile", help="Profile")
+    parser.add_argument("--endpoint-url", help="Custom S3 endpoint URL")
```
Kill all of these args

suricactus commented on scripts/object_storage_cleaner/storage_cleaner.py:
```diff
@@ -0,0 +1,639 @@
+    # Action flags
+    parser.add_argument(
+        "--info", action="store_true", help="Show bucket information (default behavior)"
+    )
```
Kill it

suricactus commented on scripts/object_storage_cleaner/storage_cleaner.py:
```diff
@@ -0,0 +1,639 @@
+    parser.add_argument("--scan", action="store_true", help="Scan only")
+
+    parser.add_argument(
+        "--prune",
```
```suggestion
    parser.add_argument(
        "--permanently-delete-versions",
```

Be scary when there is a chance to destroy data. Should be very very explicit what it is going to do.

suricactus reviewed (commented):
We are moving forward, but there are more things to be removed:

- please do not use inline ifs (binary/ternary operators) in Python code related to qfield and qfieldcloud. Python syntax for it is unreadable for people coming from other languages and it is confusing even for regular pythonists sometimes. The only reason we don't have https://github.com/afonasev/flake8-if-expr enabled is because it is slow and that one can argue they might be useful in simple situations.
- simplify logging logic: it should just log simple strings with minimal dynamic part.
- avoid flushing the buffers: there is PYTHONUNBUFFERED for people who really need it
- print everything line by line to stdout and leave the caller to do whatever they want with the buffer. It must be easy for grepping.
- do not add functionality that is very well handled by other tools, e.g. log rotation.
- avoid complex logic with logging and if statements when log appears. That's why different log levels exist. It should be very simple to follow what is going on from the log without any evaluation of if statements.
- please remove the "Connection options" params and use envvars for that purpose.
- avoid using tuple as return values, unless well argumented and documented why there is no other option.
- avoid using default variables for scripts. If you need a parameter, they should be explicit. If you don't need it explicit, then drop the parameter.
- in scripts like this do not use inline `for`s and `map`, `filter` and `reduce`. Functional code is less readable in case of emergency. 

Once addressed, please consider the following:

- remove the OOP abstraction for state management and using pure imperative code.
- avoid using the non-native container object `VersionRef`. 

The short summary is: start small and only add functionality if really required.

suricactus commented:
I am closing this, unless we want to keep it open for some reason?
